### PR TITLE
Execution retry and timeout of lifecycle hooks

### PR DESCRIPTION
**Pull request by PengTaoWW**
**What this PR does / why we need it**:
I have an improvement on Kubernetes, in our production environment, run perstop hook when removing pod, but no matter how quickly the result of this pod is removed, my improvement is that when Perstop fails, it will retry until the execution s
ucceeds, or the timeout is removed. I have implemented a version of this feature myself. if there's a problem with my improvement, please let me know, thank you.
and this time i clear code .just my patch not auto-generated code 
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:
please help me ,this is my first submit code .
**Release note**:

```release-note
containers:
  - name: lifecycle
    image: busybox
    lifecycle:
      postStart:
        exec:
          command:
            - "touch"
            - "/var/log/lifecycle/post-start"
      preStop:
        httpGet:
          path: "/abort"
          port: 8080
        isRetryUntilSuccess: true
 This PR adds a switch to control whether Prestop runs multiple times until it succeeds, or the timeout exits
The handler adds a bool value switch, which is similar to the Exec, HTTP, and Tcpsocket. But after these three switches,
```

### CODE REVIEW

**Comment by PengTaoWW** [comment]
@timstclair @brendandburns Is there anything else I need to solve?

File: api/swagger-spec/apps_v1alpha1.json
Comment (xiangpengzhao) [inline api/swagger-spec/apps_v1alpha1.json]: s/PreStop/RetryUntilSuccess

File: pkg/kubelet/kuberuntime/kuberuntime_container.go
Comment (xiangpengzhao) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: s/preStop/retryUntilSuccess

File: pkg/kubelet/kuberuntime/kuberuntime_container.go
Comment (xiangpengzhao) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: What if the hook never succeeds? The container will get stuck, right?

**Review by xiangpengzhao — CHANGES_REQUESTED** [review]

File: pkg/kubelet/kuberuntime/kuberuntime_container.go
Comment (PengTaoWW) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: @xiangpengzhao this line code will run in after long time . not stuck

File: pkg/kubelet/kuberuntime/kuberuntime_container.go
Comment (PengTaoWW) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: after some times, code line 505 will run .not get stuck

File: pkg/kubelet/kuberuntime/kuberuntime_container.go
Comment (PengTaoWW) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: i wil fixed this

File: api/swagger-spec/apps_v1alpha1.json
Comment (PengTaoWW) [inline api/swagger-spec/apps_v1alpha1.json]: my bad ,i will fixed this

File: pkg/kubelet/kuberuntime/kuberuntime_container.go
Comment (PengTaoWW) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: code line 505 will run , never get stuck

**Comment by brendandburns** [comment]
Hrm,
Rather than adding a brand new lifecycle event, why not add a boolean to enable this behavior for the existing pre-stop handler. (default false) That seems much cleaner.

**Comment by brendandburns** [comment]
(oops, didn't mean to close...)

**Comment by PengTaoWW** [comment]
@brendandburns So you mean 
containers:
  - name: lifecycle
    image: busybox
    lifecycle:
      postStart:
        exec:
          command:
            - "touch"
            - "/var/log/lifecycle/post-start"
      preStop:
        isRetryUntilSuccess: True
        httpGet:
          path: "/abort"
          port: 8080
like this ?

**Comment by PengTaoWW** [comment]
@brendandburns  i write a version code for add a boolean to enable this behavior for the existing pre-stop handler. 
@xiangpengzhao old version code will never stop ,i fix it. and this version can stop.thank you for your help

**Comment by PengTaoWW** [comment]
@brendandburns @xiangpengzhao

File: pkg/kubelet/kuberuntime/kuberuntime_container.go
Comment (brendandburns) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: This is hacky. it should just be:

```
for {
....
}
```

File: pkg/kubelet/kuberuntime/kuberuntime_container.go
Comment (brendandburns) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: Don't you want to have a `break` here?

**Comment by timstclair** [comment]
@k8s-bot ok to test

File: pkg/kubelet/kuberuntime/kuberuntime_container.go
Comment (timstclair) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: nit: I think this should be V(4) (https://github.com/kubernetes/community/blob/master/contributors/devel/logging.md)

File: pkg/kubelet/kuberuntime/kuberuntime_container.go
Comment (timstclair) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: gracePeriod should be a `time.Duration`. Same for the return type.

File: pkg/kubelet/kuberuntime/kuberuntime_container.go
Comment (timstclair) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: nit: use `time.Now()`, since this is internal.

File: pkg/kubelet/kuberuntime/kuberuntime_container.go
Comment (timstclair) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: I think this should be rewritten using `wait.Until` (https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go#L51), and set stopCh to your `time.After` timeout.

**Review by timstclair — COMMENTED** [review]
I think there's a bit of API code generation missing. See https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#changing-versioned-apis.

Also, please fill in the release note section of the PR, describing the api change.

/cc @kubernetes/api-reviewers

**Comment by fejta** [comment]
@k8s-bot pull-kubernetes-e2e-kops-aws test this
ref: kubernetes/test-infra#2932

File: api/openapi-spec/swagger.json
Comment (xiangpengzhao) [inline api/openapi-spec/swagger.json]: why adding a comma here? It's invalid. CI fails due to this.

Comment (xiangpengzhao) [inline api/openapi-spec/swagger.json]: Other generated files don't need this comma also.

File: pkg/kubelet/kuberuntime/kuberuntime_container.go
Comment (PengTaoWW) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: @brendandburns Hi , Imagine if Prestop will never succeed. The But long time timeout will trigger the select in the code, and then the container will be forcibly removed, but go func () would never stop because the container would not be able to perform the Prestop hook, so I used the For Range loop here. If there is a better plan, please let me know

File: pkg/api/types.go
Comment (xiangpengzhao) [inline pkg/api/types.go]: delete this space line, please.

File: pkg/kubelet/kuberuntime/kuberuntime_container.go
Comment (xiangpengzhao) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: I'm not sure about what if the retry time is greater than gracePeriod.

File: pkg/kubelet/kuberuntime/kuberuntime_container.go
Comment (xiangpengzhao) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: fix gofmt, see: https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46426/pull-kubernetes-verify/33233/

**Review by xiangpengzhao — CHANGES_REQUESTED** [review]

File: pkg/kubelet/kuberuntime/kuberuntime_container.go
Comment (PengTaoWW) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: @timstclair in executePreStopHook func gracePeriod is int64 type,and in all killcontainer func graceperiod initialized to init64, so if i change executeRetryUntilSuccessHook return type, means I need to modify all the values used in the killcontainer function to graceperiod. If this is what you need, I can modify.

File: pkg/kubelet/kuberuntime/kuberuntime_container.go
Comment (PengTaoWW) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: ok， i can fixed

File: pkg/kubelet/kuberuntime/kuberuntime_container.go
Comment (PengTaoWW) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: reason as same

File: pkg/kubelet/kuberuntime/kuberuntime_container.go
Comment (PengTaoWW) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: ok i can fix it

File: api/openapi-spec/swagger.json
Comment (PengTaoWW) [inline api/openapi-spec/swagger.json]: sorry for that ,i will fix

File: pkg/kubelet/kuberuntime/kuberuntime_container.go
Comment (PengTaoWW) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: Don't worry, the retry time is greater than graceperiod this will not happen because the graceperiod will trigger case <-time.After(time.Duration(gracePeriod) * time.Second).so will return。after this  i will write wait.Until replace this code.

**Comment by PengTaoWW** [comment]
@xiangpengzhao i fix in a new version code，please help me to see if there is any problem let me know
@timstclair 
 i dont't know https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#changing-versioned-apis. @timstclair say i miss something, but my pr just need modify pkg/api/types.go and /pkg/api/v1/types.go , so i miss what? and about PR in the Int64 change to time type, the reason has been explained above, if you insist to change to a timer, let me know，i try my best. go func change to wait.until is very good ,thank for you help.
@brendandburns 
i fix in a new version code ，please help me to see if there is any problem let me know

**Comment by PengTaoWW** [comment]
@timstclair i found i can't pass pull-kubernetes-baze, pull-kubernetes-e2e-kops-aws  ... because i miss some automatically generated code, but I must be ok with basice code ，then i can move on to fix test problem. So  I have fixed your previous question now？. Can we do the next step?

**Comment by PengTaoWW** [comment]
@timstclair  I submitted a version with automatic code generation, which is a lot less in the test results. but there are some mistakes left, and I wonder if I've missed some of the files. If so, please let me know.

**Comment by smarterclayton** [comment]
We are past code freeze unfortunately - this will have to wait for master to reopen for 1.8 before we can merge it.

Note that we don't use "is" as a prefix for Boolean fields in the API - we try to keep those names focused on the intent.  In this case, to be consistent with other fields, it might be better for this to be "retryOnFailure", to align with the restart policy.  Also, we try to avoid Booleans in the API when they can have multiple possible values (see RestartPolicy as an example).  I think that is broadly in line with what your are trying to do here - other policies we might want to introduce in the future are "Always" (run the hook repeatedly until the timeout) or "OnSuccess".  So it may be better to have "retryPolicy" on the hook with a default value "Never" and your feature being "OnFailure"

**Comment by PengTaoWW** [comment]
@smarterclayton Excuse me, you said the function is already developed, but did not merge into the trunk or the future planning to do so? I don't know what I need to do next, or is my work done?

**Comment by smarterclayton** [comment]
Thursday was code freeze for kubernetes 1.7, meaning no new features can be
delivered until we reopen for 1.8 features in a few weeks.  In the
meantime, review may take a while because people are fixing bugs and
getting ready to release Kube 1.7.

**Comment by PengTaoWW** [comment]
@smarterclayton I get it, version 1.8 I can submit the code again. So do I need to submit it again in 1.8? Or what else do I need to do?

**Comment by xiangpengzhao** [comment]
@PengTaoWW You don't need to submit the code again. You can still keep this PR and polish it according to folks' review suggestions during code freeze. Once master reopens for 1.8 and your code is ready to merge, it can be merged.

**Comment by PengTaoWW** [comment]
@xiangpengzhao thank you for let me know about this. 
@timstclair  If I need to modify some code to fit into the trunk, please let me know
@brendandburns If I need to modify some code to fit into the trunk, please let me know
@smarterclayton  after this version code merge into trunk. i will think about retryPolicy ,I found a question about the retry mechanism, you said that in fact the main case on the three kinds, one is never one is Onfail, but Onsuccess I think there is no need. These three retry mechanisms simply describe the scenario where the retry occurred, but for a retry strategy, such as how many retries? no.  I thought the whole Lifecle should be moving in the direction of Initcontainer. Based on these considerations I think the mechanism of using the switch to solve the long time retry in the short term will be quicker, and then hope to discuss with the Community friends about the future direction of the Lifecle. So please let me know what I can do

**Comment by smarterclayton** [comment]
We generally never add APIs that don't fit into our patterns, especially to
long term objects like Pods that we have lifecycle on.  So we would
probably not do a "quick and dirty" feature if we don't know how it will
evolve.

**Comment by PengTaoWW** [comment]
@smarterclayton ok , i write a version about retryPolicy and hope you can help code review

**Comment by brendandburns** [comment]
@PengTaoWW 

Basically just turn that boolean into a string. (and then `type string HookRetryPolicy` or somesuch)

Even though this won't merge til 1.8, I'm happy to keep reviewing in the meantime.

**Comment by PengTaoWW** [comment]
@smarterclayton  So you mean

containers:
  - name: lifecycle
    image: busybox
    lifecycle:
      postStart:
        retryPolicy: Always, Never, onFail
        exec:
          command:
            - "touch"
            - "/var/log/lifecycle/post-start"
      preStop:
        retryPolicy: Always, Never, onFail
        httpGet:
          path: "/abort"
          port: 8080
like this ? if not please let me know

@brendandburns Thank you for your advice.

**Comment by PengTaoWW** [comment]
@smarterclayton I submitted a version of the code. Retrypolicy picks a value from three values (Always, OnFail, Never) as a retry strategy. Always as strategy will retry until timeout no matter what happen. OnFail means first time fail and will retry until the timeout or retry until success. Never is native logic, only run once.please help me see it
 @brendandburns  I submitted a version of the code like you say. but i have a problem.this version code fail most tests. look like no matter with code logic. please help me let me know i miss what？
@xiangpengzhao    I submitted a version of the code  but i have a problem.this version code fail most tests. look like no matter with code logic. please help me let me know i miss what？
@timstclair  I submitted a version of the code  but i have a problem.this version code fail most tests. look like no matter with code logic. please help me let me know i miss what？
and anything if I need to do it please let me know， thank you every one

**Comment by xiangpengzhao** [comment]
> I submitted a version of the code but i have a problem.this version code fail most tests. look like no matter with code logic. please help me let me know i miss what？

Since this PR breaks API, you should also change the validation and something else related.

**Comment by PengTaoWW** [comment]
@xiangpengzhao ok ,but like this err 
```
error during ./hack/e2e-internal/e2e-up.sh: exit status 2
```
and
``` 
Error lines from build-log.txt
```
i don't know how to fix

**Comment by xiangpengzhao** [comment]
That might be test flake, you can retest to see if it still exists.

**Comment by PengTaoWW** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by dhilipkumars** [comment]
@smarterclayton @brendandburns When would we need the policy to be 'Always' for PreStop hook.  Also for PostStart should we not just stick with 'OnFail' and 'Never', otherwise the PostStart hook will never finish.

When we are re-trying pre-stop on failure, and if the prestop hook is HTTP should we not wait a bit like 100ms between the retry attempts? Otherwise, we could potentially keep AppContainer busy just dealing with the prestop hook.

**Comment by brendandburns** [comment]
Timeout between HTTP hooks makes sense (or better yet a `Retry-After` header)

Also, I agree `Always` doesn't really make sense (and the implementation is broken also as far as I can tell)

**Comment by brendandburns** [comment]
oops, didn't mean to close, apologies!

**Comment by PengTaoWW** [comment]
@brendandburns Do I need to make any changes? Or just fix test ？

**Comment by PengTaoWW** [comment]
@k8s-bot pull-kubernetes-e2e-kops-aws test this

**Comment by xiangpengzhao** [comment]
@PengTaoWW I suggest you setup pre-commit. It will run verify scripts and suggest how to do if the verification fails.

**Comment by PengTaoWW** [comment]
@xiangpengzhao ok i try

**Comment by PengTaoWW** [comment]
@xiangpengzhao i pass pre-commit but still fail

**Comment by xiangpengzhao** [comment]
check the test failure details and fix the errors please. I see some errors are related to your code.

**Comment by dhilipkumars** [comment]
@PengTaoWW I think we may have to wait a bit for Clayton to respond if we really need `Always` as a restart policy as the other approver is not in favor of it.  Also i cant find code related to PostStart hooks implementing RestartPolicy.

Where is the policy value getting defaulted the default should be `Never`  to retain backward compatibility.  Could you please split your commit into two 
1) Generated Code
2) Your implementation

**Comment by PengTaoWW** [comment]
@dhilipkumars I modified the default.go but I don't know what you mean by default, API description?
 Always my opinion is still run no matter what the results is（eg: AB test. they can still check they flow just in a hour. after this time check will stop. ）,  but there is a time slice in the program, and retry will stop depending on the time slice. You said two versions that I submitted before. But @timstclair asked me to submit the full code. So I submitted the full version to pass test.But you're right@brendandburns 、@smarterclayton 、@timstclair  What do you think?

**Comment by PengTaoWW** [comment]
@brendandburns 、@timstclair Now my code version cannot rebase the latest master code. So I can open a new PR in the latest version of this part of the work?

**Comment by xiangpengzhao** [comment]
It might not be so good an idea to send a new PR. It makes it hard to find out the current disscusion here in the future. What is the reason for not being able to rebase? If you encounter code confliction when rebasing, you should solve the confliction and then continue.

**Comment by PengTaoWW** [comment]
@xiangpengzhao ok i found clean code version rebase.

File: pkg/kubelet/kuberuntime/kuberuntime_container.go
Comment (timstclair) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: Actually, it looks like you don't need the done channel, so I think `wait.Poll` would be a better fit.

**Review by timstclair — COMMENTED** [review]
@kubernetes/sig-node-feature-requests - There was a related proposal that came up in a recent sig-node meeting for adding "FinalContainers" that act like init-containers, but for the pre-stop case. The proposals don't conflict, but just FYI since they're related.

**Comment by PengTaoWW** [comment]
@timstclair  Thank you for your code review. I found that most of the tests I could pass, but in Pull Kubernes-Verny would fail, I don ' t know how to fix. On my computer, the test is mostly through

**Comment by xiangpengzhao** [comment]
@PengTaoWW See the pull-kubernetes-verify log https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46426/pull-kubernetes-verify/36972/ and run the verification scripts which failed in it.

**Comment by PengTaoWW** [comment]
@xiangpengzhao For the present error, I do not know how to do, please help me
@timstclair For the present error, I do not know how to do, please help me

File: pkg/kubelet/kuberuntime/kuberuntime_container.go
Comment (PengTaoWW) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: @timstclair i use wati.Poll,but not use  gracePeriod  time.Duration

File: pkg/kubelet/kuberuntime/kuberuntime_container.go
Comment (PengTaoWW) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: ok thx

File: pkg/kubelet/kuberuntime/kuberuntime_container.go
Comment (PengTaoWW) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: ok i fix

**Comment by PengTaoWW** [comment]
@timstclair @brendandburns  @xiangpengzhao Is this version all right?

**Comment by PengTaoWW** [comment]
@timstclair @xiangpengzhao @brendandburns sorry about  I lost the version of my local clean code, and now I have only the version after the code is generated. What am I going to do?

**Comment by PengTaoWW** [comment]
@timstclair @xiangpengzhao @brendandburns sorry about I lost the version of my local clean code, and now I have only the version after the code is generated. What am I going to do?

**Comment by PengTaoWW** [comment]
@timstclair @xiangpengzhao @brendandburns sorry about I lost the version of my local clean code, and now I have only the version after the code is generated. What am I going to do?

**Pull request closed without merging**
